### PR TITLE
Hex package diff lock package

### PR DESCRIPTION
**Pull request by RyanSiu1995**
Issue #632. 
Implement the diff command with package inside ./deps.

### CODE REVIEW

File: lib/mix/tasks/hex.package.ex [hunk 1]
```diff
@@ -148,11 +154,48 @@ defmodule Mix.Tasks.Hex.Package do
...
+    locked_deps = Mix.Dep.Lock.read()
+
+    path_lock =
+      case Map.get(locked_deps, String.to_atom(package)) do
+        nil ->
+          Mix.raise("Expected packages defined in lockfile, got: `\"#{package}\"`")
+
+        _ ->
+          unescaped_path = Path.join(Mix.Project.deps_path(), package)
+          String.replace(unescaped_path, "\"", "\\\"")
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex]: why do we need this?

File: lib/mix/tasks/hex.package.ex [hunk 2]
```diff
@@ -148,11 +154,48 @@ defmodule Mix.Tasks.Hex.Package do
...
+
+      %{checksum: checksum} = Hex.unpack_tar!({:binary, tarball}, path)
+      verify_tar_checksum!(repo, package, version, checksum)
+
+      Hex.Registry.Server.close()
+
+      cmd =
+        Hex.State.fetch!(:diff_command)
+        |> String.replace("__PATH1__", "\"#{path_lock}\"")
+        |> String.replace("__PATH2__", path)
```
Comment (wojtekmach) [inline lib/mix/tasks/hex.package.ex]: instead of duplicating this logic, please extract common parts into a private function.

File: test/mix/tasks/hex.package_test.exs [hunk 3]
```diff
@@ -18,7 +18,34 @@ defmodule Mix.Tasks.Hex.PackageTest do
...
+
+      assert_received {:mix_shell, :run, [out]}
+      assert out =~ ~s(-{<<"version">>,<<"0.0.1">>}.)
+      assert out =~ ~s(+{<<"version">>,<<"0.1.0">>}.)
+    end)
+  after
+    purge([ReleaseDeps.MixProject])
+  end
+
+  test "diff: not existed package in lockfile" do
```
Comment (wojtekmach) [inline test/mix/tasks/hex.package_test.exs]: Thanks for adding this test. I believe we need to test these 2 edge cases:

1. We run the command in a directory that is not a Mix project (no mix.exs, no mix.lock). This is this test. I believe in this case we should tell the user that this variant of the command can only be used in Mix projects
2. We run the command in a Mix project that does not have a lockfile or has outdated lock (user needs to call mix deps.get)

File: lib/mix/tasks/hex.package.ex (hunk 1 above)
Comment (RyanSiu1995) [reply lib/mix/tasks/hex.package.ex]: The project path from user may include space character.
So, the diff command will be 
```
git diff /to/folder/with space /var/sth
```
I have to enclose that path with quote so I need to escape it before generating the command

**Comment by RyanSiu1995** [comment]
I am going to open another pull request on fixing the build.

**Comment by RyanSiu1995** [comment]
@wojtekmach Why the dependency isn't locked after running the `Mix.Tasks.Hex.Get.run([])` under elixir 1.1 and 1.0?
I cannot identify the root cause...

**Comment by ericmj** [comment]
It is possible it is a bug in those elixir versions. Could you write the lockfile manually to work around the issue?

**Comment by ericmj** [comment]
@RyanSiu1995 Ping! Any updates to this PR?

**Comment by RyanSiu1995** [comment]
Hey @ericmj, sorry for missing this.
I have rebased it to master.
But unfortunately, I still cannot resolve the issue in the legacy elixir version.
Therefore, I have just skipped those tests first.
We may fix it later on if you don't mind.

**Comment by ericmj** [comment]
I have pushed some changes to this PR, the major things are:

* Fixed failing tests on Elixir <1.2
* Differentiate between when an app name or package name is used
* Cleaned up the tests so that there are no root files in the root project directory

Big thanks for your work @RyanSiu1995. I think this is good to go, we just need another review from a core member.

**Review by ericmj — APPROVED** [review]

**Comment by RyanSiu1995** [comment]
@wojtekmach could you mind to have a final review on this?

**Comment by wojtekmach** [comment]
Looks good to me, thank you! We've since renamed our default branch from master to main and this PR was opened against the former, when I switched the base branch we got some conflicts. I'll try to solve them soon but if anyone wants to jump in in the meantime, feel free to!

**Comment by RyanSiu1995** [comment]
After merging to main, I found that all tests were failed. I will fix it later today.

**Comment by ericmj** [comment]
It looks like the merge also reverted some changes in https://github.com/hexpm/hex/pull/705/commits/48bddfd7ac72b5c71cf8a690f3d690987886d72c.

**Comment by RyanSiu1995** [comment]
@ericmj My local setting has some issues. I can't do the development with my machine right now.
Could you mind to help me to fix it?

**Comment by xinz** [comment]
Hello, please review https://github.com/hexpm/hex/pull/819 to resolve the conflict, and make the tests passed.

**Comment by wojtekmach** [comment]
Closing in favour of #819
